### ISSUE TITLE
static analyzer false positive:  unary minus

### ISSUE THREAD

**Issue by buczek** [llvm/llvm-project#55241]
This code

```
int something(void);
void out(int value);

void test() {

    int value;

    int error = something();

    if (error == 0)
        value = something();

    if (-error == 0)
        out(value);
}
```

triggers `test.c:14:9: warning: 1st function call argument is an uninitialized value [core.CallAndMessage]` with llvm 13.

The code might run into UB if `error==INT_MIN` because of the integer overflow in the unary minus. Still the message is misleading. Also, neither an assert() nor an assume-macro  with `error >= 0` after ` int error = something();` makes the bug report go away.

Scanner Output:
```
scan-build: Using '/pkg/llvm-13.0.1-0/bin/clang-13' for static analysis
scan-build: Emitting reports for this run to '/scratch/local/scan-build-2022-05-03-110530-10997-1'.

[LOCATION]: /home/buczek/scan_build_problem
'/pkg/llvm-13.0.1-0/bin/clang-13' '-cc1' '-triple' 'x86_64-unknown-linux-gnu' '-analyze' '-disable-free' '-main-file-name' 'test.c' '-analyzer-store=region' '-analyzer-opt-analyze-nested-blocks' '-analyzer-checker=core' '-analyzer-checker=apiModeling' '-analyzer-checker=unix' '-analyzer-checker=deadcode' '-analyzer-checker=security.insecureAPI.UncheckedReturn' '-analyzer-checker=security.insecureAPI.getpw' '-analyzer-checker=security.insecureAPI.gets' '-analyzer-checker=security.insecureAPI.mktemp' '-analyzer-checker=security.insecureAPI.mkstemp' '-analyzer-checker=security.insecureAPI.vfork' '-analyzer-checker=nullability.NullPassedToNonnull' '-analyzer-checker=nullability.NullReturnedFromNonnull' '-analyzer-output' 'plist' '-w' '-setup-static-analyzer' '-mrelocation-model' 'static' '-mframe-pointer=all' '-fmath-errno' '-fno-rounding-math' '-mconstructor-aliases' '-munwind-tables' '-target-cpu' 'x86-64' '-tune-cpu' 'generic' '-debugger-tuning=gdb' '-fcoverage-compilation-dir=/home/buczek/scan_build_problem' '-resource-dir' '/pkg/llvm-13.0.1-0/lib/clang/13.0.1' '-internal-isystem' '/pkg/llvm-13.0.1-0/lib/clang/13.0.1/include' '-internal-isystem' '/usr/local/include' '-internal-isystem' '/usr/lib64/gcc/x86_64-pc-linux-gnu/7.5.0/../../../../x86_64-pc-linux-gnu/include' '-internal-externc-isystem' '/include' '-internal-externc-isystem' '/usr/include' '-fdebug-compilation-dir=/home/buczek/scan_build_problem' '-ferror-limit' '19' '-fgnuc-version=4.2.1' '-analyzer-display-progress' '-analyzer-output=html' '-faddrsig' '-D__GCC_HAVE_DWARF2_CFI_ASM=1' '-o' '/scratch/local/scan-build-2022-05-03-110530-10997-1' '-x' 'c' 'test.c'
ANALYZE (Syntax): test.c test : 0.1 ms
ANALYZE (Path,  Inline_Regular): test.c test : 1.9 ms
test.c:18:9: warning: 1st function call argument is an uninitialized value [core.CallAndMessage]
        out(value);
        ^~~~~~~~~~
===-------------------------------------------------------------------------===
                                Analyzer timers
===-------------------------------------------------------------------------===
  Total Execution Time: 0.0018 seconds (0.0024 wall clock)

   ---User Time---   --System Time--   --User+System--   ---Wall Time---  --- Name ---
   0.0014 ( 78.2%)   0.0000 (  0.0%)   0.0014 ( 74.5%)   0.0019 ( 80.2%)  Path exploration time
   0.0003 ( 18.9%)   0.0000 (  0.0%)   0.0003 ( 18.0%)   0.0003 ( 13.9%)  Path-sensitive report post-processing time
   0.0000 (  2.8%)   0.0001 (100.0%)   0.0001 (  7.5%)   0.0001 (  5.9%)  Syntax-based analysis time
   0.0017 (100.0%)   0.0001 (100.0%)   0.0018 (100.0%)   0.0024 (100.0%)  Total

1 warning generated.
scan-build: Analysis run complete.
scan-build: 1 bug found.
scan-build: Run 'scan-view /scratch/local/scan-build-2022-05-03-110530-10997-1' to examine bug reports.
```

**Comment by steakhal** [comment]
It can be reduced to:
```c
int test(int flag) {
  int value;
  if (flag == 0)
    value = 1;
  if (-flag == 0)
    return value;
  return 42;
}
```
`ExprEngine::Visit()` handles the `case Stmt::UnaryOperatorClass`.
Which will call `ExprEngine::VisitUnaryOperator()`, which handles the `case UO_Minus`.
That will call `evalMinus(V.castAs<NonLoc>())` of the `SimpleSValBuilder` class.
It turns out it's only implemented for `ConcreteInt`s, but it seems to be difficult to generalize this.

```c++
SVal SimpleSValBuilder::evalMinus(NonLoc val) {
  switch (val.getSubKind()) {
  case nonloc::ConcreteIntKind:
    return val.castAs<nonloc::ConcreteInt>().evalMinus(*this);
  default:
    return UnknownVal(); // <--
  }
}
```

Maybe the new solver could figure out something if we would conjure a new symbol and add a constraint that it's equal to the negated original value. @martong WDYT?

**Comment by haoNoQ** [comment]
Making a new symbol class for unary expressions such as unary minus is obviously the most principled solution. Bitwise negation will also fall into the same category eventually.

**Comment by buczek** [comment]
Would this possible solution cover this more complex example? This is relevant for functions converting between errno-style api to return-negative-on-failure api. This requires the knowledge that if `i > 0` (maybe from assert or assume) then `-i < 0`.

```c
#include <assert.h>
int something(void);
void do_something_with(int value);
int error;

int bar(int *value) {
  if (something() == -1) {
    assert(error > 0);
    return -error;
  }
  *value = 123;
  return 0;
}

void test(void) {
    int value;
    int res;
    res = bar(&value);
    if (res < 0)
      return;
    do_something_with(value);
}
```

**Comment by martong** [comment]
> It can be reduced to:
> 
> ```c
> int test(int flag) {
>   int value;
>   if (flag == 0)
>     value = 1;
>   if (-flag == 0)
>     return value;
>   return 42;
> }
> ```
> 
> `ExprEngine::Visit()` handles the `case Stmt::UnaryOperatorClass`. Which will call `ExprEngine::VisitUnaryOperator()`, which handles the `case UO_Minus`. That will call `evalMinus(V.castAs<NonLoc>())` of the `SimpleSValBuilder` class. It turns out it's only implemented for `ConcreteInt`s, but it seems to be difficult to generalize this.
> 
> ```c++
> SVal SimpleSValBuilder::evalMinus(NonLoc val) {
>   switch (val.getSubKind()) {
>   case nonloc::ConcreteIntKind:
>     return val.castAs<nonloc::ConcreteInt>().evalMinus(*this);
>   default:
>     return UnknownVal(); // <--
>   }
> }
> ```
> 
> Maybe the new solver could figure out something if we would conjure a new symbol and add a constraint that it's equal to the negated original value. @martong WDYT?

We could handle the special case of assigning `0` in the ConstraintAssignor. I.e. we would be able to create the negated symbol and assign the very same constraint for that. 
However, to have the generic solution we should have a support of negating a `RangeSet`, which seems quite convoluted.

**Comment by martong** [comment]
> However, to have the generic solution we should have a support of negating a RangeSet, which seems quite convoluted.
Perhaps, we might be able to reuse some parts of the logic of the signedness conversion we already have in `castTo`. @ASDenysPetrov WDYT?

**Comment by martong** [comment]
> Making a new symbol class for unary expressions such as unary minus is obviously the most principled solution. Bitwise negation will also fall into the same category eventually.

Yes, that would be the  generic solution. And in `SValBuilder` it would reach down to the `negate` function in the ConstraintSolver if we had it.

**Comment by martong** [comment]
First patch is on the way:
https://reviews.llvm.org/D125318 [analyzer] Add UnarySymExpr

A subsequent patch might add the missing logic to the ConstraintSolver.

**Comment by ASDenysPetrov** [comment]
> Perhaps, we might be able to reuse some parts of the logic of the signedness conversion we already have in `castTo`. @ASDenysPetrov WDYT?
I'm not sure I got your point. but AFAIK We already have `negate` function for ranges.
> We could handle the special case of assigning 0 in the ConstraintAssignor.
Beware. First, unary minus obeys [[conv.prom]](https://eel.is/c++draft/conv.prom). Second, in practice, every _middle_ number is a special case in integer types, e.g. 128 of uchar, 2147483648 of uint are also a _special cases_ that turn into the same number like 0.

**Comment by martong** [comment]
> > Perhaps, we might be able to reuse some parts of the logic of the signedness conversion we already have in `castTo`. @ASDenysPetrov WDYT?
> 
> I'm not sure I got your point. however we already have `negate` function for ranges.
> 
> > We could handle the special case of assigning 0 in the ConstraintAssignor.
> 
> Beware. First, unary minus obeys [[conv.prom]](https://eel.is/c++draft/conv.prom). Second, in practice, every _first_ and _middle_ numbers are _special cases_ in integer types, e.g. -128 of char, 128 of uchar, -2147483648 of int, 2147483648 of uint turn into the same number. That's why 0 turns into 0 in every type, because it is either on the first place or in the middle of the integer range.

Yes, so true. Fortunately, we went on with the more generic solution by introducing the `UnarySymExpr`.
